akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRecoveringEvents.scala

```diff
@@ -0,0 +1,206 @@
+        state = newState
+        same
+
+      case RecoverySuccess(highestSeqNr) ⇒
+        log.debug("Recovery successful, recovered until sequenceNr: {}", highestSeqNr)
+        cancelRecoveryTimer()
+        setLastSequenceNr(highestSeqNr)
+
+        try onRecoveryCompleted(state)
+        catch { case NonFatal(ex) ⇒ onRecoveryFailure(ex, Some(state)) }
+
+      case ReplayMessagesFailure(cause) ⇒
+        onRecoveryFailure(cause, event = None)
+
+      case other ⇒
+        stash(other, same)
+    }
+  } catch {
+    case NonFatal(e) ⇒
+      cancelRecoveryTimer()
+      onRecoveryFailure(e, None)
+  }
+
+  def onSnapshotterResponse(response: SnapshotProtocol.Response): Behavior[Any] = {
+    log.warning("Unexpected [{}] from SnapshotStore, already in recovering events state.", Logging.simpleName(response))
+    same // ignore
+  }
+
+  /**
+   * Called whenever a message replay fails. By default it logs the error.
+   *
+   * The actor is always stopped after this method has been invoked.
+   *
+   * @param cause failure cause.
+   * @param event the event that was processed in `receiveRecover`, if the exception was thrown there
+   */
+  protected def onRecoveryFailure(cause: Throwable, event: Option[Any]): Behavior[Any] = {
+    returnRecoveryPermit("on recovery failure: " + cause.getMessage)
+    cancelRecoveryTimer()
+
+    event match {
+      case Some(evt) ⇒
+        log.error(cause, "Exception in receiveRecover when replaying event type [{}] with sequence number [{}].", evt.getClass.getName, lastSequenceNr, persistenceId)
+        Behaviors.stopped
+
+      case None ⇒
+        log.error(cause, "Persistence failure when replaying events for persistenceId [{}]. " +
+          "Last known sequence number [{}]", persistenceId, lastSequenceNr)
+        Behaviors.stopped
+    }
+  }
+
+  protected def onRecoveryCompleted(state: State): Behavior[Any] = {
+    try {
+      returnRecoveryPermit("finally in on recovery completed")
+      recoveryCompleted(commandContext, state)
+
+      val b = this
+      val running = new EventsourcedRunning[Command, Event, State](context, internalStash, _lastSequenceNr, writerIdentity) {
+        override def timers: TimerScheduler[Any] = b.timers
+        override def persistenceId: String = b.persistenceId
+        override def initialState: State = state
+        override def commandHandler: CommandHandler[Command, Event, State] = b.commandHandler
+        override def eventHandler: (State, Event) ⇒ State = b.eventHandler
+        override def recoveryCompleted: (ActorContext[Command], State) ⇒ Unit = b.recoveryCompleted
+        override def snapshotWhen: (State, Event, SeqNr) ⇒ Boolean = b.snapshotWhen
+        override def tagger: Event ⇒ Set[String] = b.tagger
+        override def journalPluginId: String = b.journalPluginId
+        override def snapshotPluginId: String = b.snapshotPluginId
+      }
+
+      tryUnstash(context, running)
```

Comment:
ok, so this is unstashing one command, and then additional unstashing will be triggered by the returned effects?

it's also unstashing in onWriteMessageComplete - not sure about the call order but make sure that the effects are run first and then the unstashing, so that an unstashed command is not run before the side effects of previous